src/Features/CSharp/Portable/ConvertLinq/CSharpConvertLinqQueryToForEachProvider.cs

```diff
@@ -0,0 +1,792 @@
+                    }
+                }
+
+                if (TryConvertQueryExpression(source, internalNodeMethod, out StatementSyntax newNode))
+                {
+                    documentUpdateInfo = new DocumentUpdateInfo(forEachStatement, newNode.WithAdditionalAnnotations(Simplifier.Annotation));
+                    return true;
+                }
+
+                documentUpdateInfo = default;
+                return false;
+            }
+
+            private bool TryConvertIfInReturnStatement(QueryExpressionSyntax source, ReturnStatementSyntax returnStatement, out DocumentUpdateInfo documentUpdateInfo)
+            {
+                // before: return from a in b select a;
+                // after: 
+                // foreach(var a in b)
+                // {
+                //      yield return a;
+                // }
+                StatementSyntax internalNodeMethod(ExpressionSyntax expression)
+                    => SyntaxFactory.YieldStatement(SyntaxKind.YieldReturnStatement, expression);
+
+                if (TryConvertQueryExpression(source, internalNodeMethod, out var newNode))
+                {
+                    documentUpdateInfo = new DocumentUpdateInfo(returnStatement, newNode.WithAdditionalAnnotations(Simplifier.Annotation));
+                    return true;
+                }
+
+                documentUpdateInfo = default;
+                return false;
+            }
+
+            private bool TryConvertQueryExpression(
+                QueryExpressionSyntax source,
+                Func<ExpressionSyntax, StatementSyntax> leafExpressionCreationMethod,
+                out StatementSyntax statement)
+            {
+                ProcessingState processingState = new ProcessingState(leafExpressionCreationMethod);
+                if (TryProcessFromClause(source.FromClause, processingState) &&
+                    TryProcessQueryBody(source.Body, processingState))
+                {
+                    // Executes syntax building methods from bottom to the top of the tree.
+                    statement = processingState.ExecuteMethods().WithAdditionalAnnotations(Simplifier.Annotation);
+                    return true;
+                }
+
+                statement = default;
+                return false;
+            }
+
+            private bool TryProcessQueryBody(
+                QueryBodySyntax queryBody,
+                ProcessingState processingState)
+            {
+                foreach (var queryClause in queryBody.Clauses)
+                {
+                    if (!TryProcessQueryClause(queryClause, processingState))
+                    {
+                        return false;
+                    }
+                }
+
+                // GroupClause is not supported by the conversion
+                return queryBody.SelectOrGroup is SelectClauseSyntax selectClause &&
```

Comment:
>queryBody.SelectOrGroup is SelectClauseSyntax selectClause [](start = 23, length = 58)

Why do we require presence of Select in order to handle Continuation?

Reply:
As far as I know, there are two continuations: 'select into' and 'group into'. We do not support 'group by' in the conversion. Therefore, continuation appears in the 'select into' context. Correct?

---
In reply to: [177187616](https://github.com/dotnet/roslyn/pull/25362#discussion_r177187616) [](ancestors = 177187616)